### PR TITLE
Handle TypedDict in diff and deps

### PR DESCRIPTION
**Pull request by msullivan**
Required performing a patchup in semanal to *actually* replace the
TypeInfo with the 'replaced' version.

### CODE REVIEW

File: mypy/server/deps.py [hunk 1]
```diff
@@ -199,6 +199,8 @@ def visit_class_def(self, o: ClassDef) -> None:
             self.add_type_dependencies(base, target=target)
         if o.info.tuple_type:
             self.add_type_dependencies(o.info.tuple_type, target=make_trigger(target))
+        if o.info.typeddict_type:
```
Comment (ilevkivskyi) [inline mypy/server/deps.py]: I am mostly in favor of "unification" of ``TypedDict`` and ``NamedTuple``, but why can't you just use `o.info.replaced.typeddict_type` here? (Then you wouldn't need to change `semanal.py`.)

File: mypy/server/deps.py [hunk 2]
```diff
@@ -258,6 +259,12 @@ def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
                         self.add_type_dependencies(typ, target=make_trigger(prefix))
                         attr_target = make_trigger('%s.%s' % (prefix, name))
                         self.add_type_dependencies(typ, target=attr_target)
+        elif isinstance(rvalue, CallExpr) and isinstance(rvalue.analyzed, TypedDictExpr):
+            # Depend on the underlying typeddict type
+            info = rvalue.analyzed.info
+            assert(info.typeddict_type is not None)
```
Comment (JukkaL) [inline mypy/server/deps.py]: Style nit: spurious parens around the expression.

File: test-data/unit/diff.test [hunk 3]
```diff
@@ -562,3 +562,33 @@ B = Enum('B', 'y')
...
+[file next.py]
+from mypy_extensions import TypedDict
+Point = TypedDict('Point', {'x': int, 'y': str})
+p = Point(dict(x=42, y='lurr'))
+[builtins fixtures/dict.pyi]
+[out]
+__main__.Point
+__main__.p
+
+[case testTypedDict2]
```
Comment (JukkaL) [inline test-data/unit/diff.test]: Test also that removing or adding a field triggers a TypedDict type. And test that changing totality triggers the type.

File: test-data/unit/fine-grained.test [hunk 4]
```diff
@@ -1669,3 +1669,59 @@ def f(x: a.N) -> None:
...
+[file b.py]
+from a import Point
+def foo(x: Point) -> int:
+    return x['x'] + x['y']
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:3: error: Unsupported operand types for + ("int" and "str")
+
+[case testTypedDictUpdate]
```
Comment (JukkaL) [inline test-data/unit/fine-grained.test]: Test case has same name as the previous test case.

File: test-data/unit/semanal-typeddict.test [hunk 5]
```diff
@@ -46,8 +46,8 @@ MypyFile:1(
   ImportFrom:1(mypy_extensions, [TypedDict])
   ClassDef:2(
     A
-    BaseTypeExpr(
-      NameExpr(TypedDict [mypy_extensions.TypedDict]))
+    BaseType(
+      typing.Mapping[builtins.str, builtins.str])
```
Comment (JukkaL) [inline test-data/unit/semanal-typeddict.test]: What's this change?

File: mypy/server/deps.py (hunk 1 above)
Comment (msullivan) [reply mypy/server/deps.py]: `replaced` doesn't get serialized, so the right thing to do seemed to be following `NamedTuple` and actually doing the replacement.

File: test-data/unit/semanal-typeddict.test (hunk 5 above)
Comment (msullivan) [reply test-data/unit/semanal-typeddict.test]: A consequence of the semanal changes that replace the TypeInfo with the `replaced` version.

**Pull request merged**